model/vfs/couchdb_indexer.go

```diff
@@ -536,14 +563,29 @@ func (c *couchdbIndexer) CheckIndexIntegrity(accumulate func(*FsckLog)) (err err
 		}
 	}
 
-	return
+	return couchdb.ForeachDocs(c.db, consts.FilesVersions, func(_ string, data json.RawMessage) error {
+		v := &Version{}
+		if erru := json.Unmarshal(data, v); erru != nil {
+			return erru
+		}
+		fileID := strings.SplitN(v.DocID, "/", 2)[0]
+		if _, ok := tree.Files[fileID]; !ok {
+			accumulate(&FsckLog{
+				Type:       FileMissing,
+				IsVersion:  true,
+				VersionDoc: v,
+			})
+		}
+		return nil
```

Comment:
I don't understand the `return foreach` containing a `return`. I would appreciate if you could take 5 minutes to teach me the rationale behind this contruction.

Reply:
It is just a way to return an error. You can see this code as a equivalent of:

```go
var docs []*Version
err := couchdb.GetAllDocs(c.db, consts.FilesVersions, &docs)
if err != nil {
    return err
}
for _, v := range docs {
    if err = doSomething(v); err != nil {
        return err
    }
}
return nil
```

But, this versions loads all the documents in memory from a single request to CouchDB. To avoid that, we are using pagination in `couchdb.ForeachDocs`. And, with the lack of generics in go, we have to call the callback function with a `json.RawMessage` and let it unmarshal the document. This can create an error, and if it is the case, we want to send back the error to the main function. That's all.